### PR TITLE
move out the regex matching code to better readability #1439

### PR DESCRIPTION
**Pull request by muhammadn**
Move out the regex filter code from Channel::Filter::Database into a separate concern with tests included

### CODE REVIEW

File: app/models/concerns/email_regex.rb [hunk 1]
```diff
@@ -0,0 +1,28 @@
+module EmailRegex
```
Comment (thorsteneckel) [inline app/models/concerns/email_regex.rb]: Two things:
- Modules should be used for mixins. This case is an isolated logic task which should be a class.
- `EmailRegex` is a to generic name for this specific use case. Please use the namespace `Channel::Filter::Match::EmailRegex` instead.

File: app/models/concerns/email_regex.rb [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+  extend ActiveSupport::Concern
```
Comment (thorsteneckel) [inline app/models/concerns/email_regex.rb]: Why are you using a concern here?

File: app/models/concerns/email_regex.rb [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+
+  def self.match(value, match_rule, _should_match, check_mode = false)
```
Comment (thorsteneckel) [inline app/models/concerns/email_regex.rb]: Two things:
- Why is there an argument `_should_match` which is never used?
- We should use keyword arguments. A rule of thumb: More than two arguments should always be keyword arguments. This increases the readability a lot.

File: test/unit/email_regex_test.rb [hunk 4]
```diff
@@ -0,0 +1,127 @@
+# encoding: utf-8
+# rubocop:disable all
```
Comment (thorsteneckel) [inline test/unit/email_regex_test.rb]: Don't disable Rubocop in new files.

File: test/unit/email_regex_test.rb [hunk 5]
```diff
@@ -0,0 +1,127 @@
...
+require 'test_helper'
+
+class EmailRegexTest < ActiveSupport::TestCase
```
Comment (thorsteneckel) [inline test/unit/email_regex_test.rb]: Just an information: If you prefer RSpec we would gladly see RSpec tests in the future over Minitest.

File: app/models/concerns/email_regex.rb (hunk 3 above)
Comment (muhammadn) [reply app/models/concerns/email_regex.rb]: yeah. i was using positional arguments and i think keyword arguments will increase readablity. _should_match came from the original method which i forgot to exclude from earlier.

File: app/models/concerns/email_regex.rb (hunk 2 above)
Comment (muhammadn) [reply app/models/concerns/email_regex.rb]: Shouldn't be there. you are right.

File: app/models/concerns/email_regex.rb (hunk 1 above)
Comment (muhammadn) [reply app/models/concerns/email_regex.rb]: Yeap. i will namespace that module. will see the new updated code soon.

File: test/unit/email_regex_test.rb (hunk 4 above)
Comment (muhammadn) [reply test/unit/email_regex_test.rb]: Sure. i am beginning to like rubocop. Never used it but it is really nice to have the code to be in order.

File: app/models/channel/filter/database.rb [hunk 6]
```diff
@@ -18,12 +18,12 @@ def self.run(_channel, mail)
           match_rule = meta['value']
           min_one_rule_exists = true
           if meta[:operator] == 'contains not'
-            if value.present? && match(value, match_rule, false)
+            if value.present? && Channel::Filter::Match::EmailRegex.match(value: value, match_rule: match_rule, check_mode: false)
```
Comment (thorsteneckel) [inline app/models/channel/filter/database.rb]: Looks like the change of attributes has introduced a behavior change. The third parameter `false` was previously the now removed `_should_match` but is now `check_mode`.

File: app/models/channel/filter/database.rb [hunk 7]
```diff
@@ -18,12 +18,12 @@ def self.run(_channel, mail)
...
               all_matches_ok = false
               Rails.logger.info "  matching #{key.downcase}:'#{value}' on #{match_rule}, but shoud not"
             end
           elsif meta[:operator] == 'contains'
-            if value.blank? || !match(value, match_rule, true)
+            if value.blank? || !Channel::Filter::Match::EmailRegex.match(value: value, match_rule: match_rule, check_mode: true)
```
Comment (thorsteneckel) [inline app/models/channel/filter/database.rb]: Looks like the change of attributes has introduced a behavior change. The third parameter `true` was previously the now removed `_should_match` but is now `check_mode`.

File: app/models/postmaster_filter.rb [hunk 8]
```diff
@@ -14,11 +14,7 @@ def validate_condition
       raise Exceptions::UnprocessableEntity, 'operator invalid, ony "contains" and "contains not" is supported' if meta['operator'].blank? || meta['operator'] !~ /^(contains|contains not)$/
       raise Exceptions::UnprocessableEntity, 'value invalid/empty' if meta['value'].blank?
       begin
-        if meta['operator'] == 'contains not'
-          Channel::Filter::Database.match('test content', meta['value'], false, true)
-        else
-          Channel::Filter::Database.match('test content', meta['value'], true, true)
-        end
+        Channel::Filter::Match::EmailRegex.match(value: 'test content', match_rule: meta['value'], check_mode: true)
```
Comment (thorsteneckel) [inline app/models/postmaster_filter.rb]: nice spot!

**Comment by thorsteneckel** [comment]
Merged with the squash commit. Thanks a lot @muhammadn 🚀

**Pull request closed without merging**